### PR TITLE
TypeScript types

### LINKED ISSUE #389: TypeScript support
After chatting with the Actions team a few months ago, I thought it might be useful to add TypeScript support to our components. @mxstbr pointed out this library that might be able to help us easily do that: https://github.com/atlassian/extract-react-types

### PR DESCRIPTION
**Pull request by mxstbr**
Adds TypeScript types to the components, so that any consumer only has to install the package and will automatically get the types. These were extracted from our internal project where I typed all the components we are currently using.

**TODO**

- [x] Currently only file-based imports are typed (`import Flex from '@primer/components/src/Flex'`), we will need to update the types to also type the exports
- [x] Avatar
- [x] BaseStyles
- [x] BorderBox
- [x] Box
- [x] BranchName
- [x] Button
- [x] CircleBadge
- [x] CircleOcticon
- [x] CounterLabel
- [x] Details
- [x] Dropdown
- [x] FilterList
- [x] Flash
- [x] Flex
- [x] Heading
- [x] Label
- [x] Link
- [x] PointerBox
- [x] Position
- [x] StateLabel
- [x] StyledOcticon
- [x] Text
- [x] TextInput
- [x] Tooltip
- [x] UnderlineNav
- [x] theme

Closes #389 

I think it might be worth having a discussion on whether the package itself should be written in TypeScript, but that would be a much bigger change and this gets us 100% of the way there for all consumers.

#### Merge checklist
- [ ] Changed base branch to release branch
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

### CODE REVIEW

**Comment by JerryGreen** [comment]
That's a cool thing! I'll subscribe to updates, will be waiting :)

P.S. Thanks for your work

**Comment by colebemis** [comment]
Really excited about this!

**Comment by mxstbr** [comment]
Done, I have now typed all components! :tada: This is ready for a review 👍

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+  export interface UnderlineNavProps extends CommonProps {
+    actions?: React.ReactNode
+    align?: 'right'
+    full?: boolean
+    label?: string
+  }
+
+  export const UnderlineNav: React.FunctionComponent<UnderlineNavProps>
+
+  export const theme: Object
```
Comment (mxstbr) [inline index.d.ts]: Note that the theme should be properly typed, but that would be a lot of effort 😅 I think this is good enough ™️ for a TypeScript MVP, and we can always improve this later.

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ Description of changes
 
 #### Merge checklist
 - [ ] Changed base branch to release branch
+- [ ] Add or update TypeScript definitions (`index.d.ts`) if necessary
```
Comment (mxstbr) [inline .github/pull_request_template.md]: I added this little reminder to ensure the TypeScript definitions stay up to date, even when new components are added or old ones changed—happy to remove that again if you think it's unnecessary

**Pull request merged**